lab-bundle-spec.md

```diff
@@ -158,21 +153,35 @@ environment_resources:
 
 ##### GCP Project (gcp_project)
 
-attribute                     | required | type    | notes
------------------------------ | -------- | ------- | --------------------------------------
-dm_template.script            |          | path    | Relative path to a Deployment Manager directory tree.
-dm_template.custom_properties |          | array   | Array of key/value pairs.
-fleet                         |          | enum*   | Specify a Qwiklabs fleet to pull the project from.
+attribute                           | required | type    | notes
+----------------------------------- | -------- | ------- | --------------------------------------
+dm_template.start.script            |          | path    | Relative path to a Deployment Manager directory tree.
+dm_template.start.custom_properties |          | array   | Array of key/value pairs.
+dm_template.end.script              |          | path    | Relative path to a Deployment Manager directory tree.
+dm_template.end.custom_properties   |          | array   | Array of key/value pairs.
+fleet                               |          | enum*   | Specify a Qwiklabs fleet to pull the project from.
+service_account_permissions         |          | array   | Array of project/roles(array) pairs to be granted to the default compute engine service account in this project.
+ssh_key_user                        |          | string  | If this project should use a user's SSH key, the id of that user.
 
 ```yml
 - type: gcp_project
   id: secondary_project
   fleet: gcpfree
   dm_template:
-    script: deployment_manager
-    custom_properties:
-      - key: userNameWindows
-        value: student
+    start:
+      script: dm_startup.zip
+      custom_properties:
+        - key: userNameWindows
+          value: student
```

Comment:
we need to figure out how to allow these values to reference properties/outputs from other resources (the authoring side part of #4081)

Reply:
ah right. I think it makes the most sense for custom_properties to have a `key` and EITHER a `value` (string that is passed through) OR `reference` (in the form `[RESOURCE].[PROPERTY]`) rather than try to smuggle both through as `value`. Does that seem reasonable to you? Also, do you have a list of the possible properties per resource?

Reply:
i think this is more of a question for @jgan. i was original in favor of that, but he had mentioned using some reference scheme like is done in deployment manager.

the reasoning behind having references was to have a replacement for the default list of inputs we pass to deployment manager (lib/ice/gcp/deploymentmanager_v2.rb#161). we may want to expand the list of properties that resources can have, but the ones currently implemented are:
- user
  - password
  - username
- project
  - project_id
  - gcp_zone (maybe should just be 'zone' to match existing dm input name)

currently not accessible through project (but passed in through default dm properties):
  - region
  - key_file_json
  - qwiklabsUuid

Reply:
I had assumed we were going to use an interpolation syntax, a la [Terraform](https://www.terraform.io/docs/configuration-0-11/interpolation.html). But, after mulling it for a bit, I'm for the more explicit distinction. It will be simpler to implement and validate.

The primary drawback I see is the lack of ability to do any value manipulation before getting passed to the script.

1) Simple value manipulation (e.g. concat)

```yaml
custom_properties:
  - key: userNameWindows
  - value: "${user.name}-pre-defenestration"
```

I'm not concerned about this, because the target scripting language can handle these value manipulations.

2) Combining an arbitrary number of values, e.g. to construct a list

```yaml
custom_properties:
  - key: listOfUsers
  - value: "[ ${user1.username}, ${user2.username}, ${user3.username} ]"
```

This only effects the scripts re-usability. The script could be re-worked to accept N different custom_properties instead of an array. I suspect this will have a larger impact on clean-up scripts than start-up scripts.

__Net-Net:__ These arguments against the explicit `reference` syntax smell like YAGNI to me. Furthermore, the reference syntax can be expanded to handle more complex usecases (e.g. lists).